refactor(compiler): support external runtime component styles for file-based stylesheets

clydin opened the pull request:
The AOT compiler now has the capability to handle component stylesheet files as external runtime files. External runtime files are stylesheets that are not embedded within the component code at build time. Instead a URL path is emitted within a component's metadata. When combined with separate updates to the shared style host and DOM renderer, this will allow these stylesheet files to be fetched and processed by a development server on-demand. This behavior is controlled by an internal compiler option `externalRuntimeStyles`. The Angular CLI development server will also be updated to provide the serving functionality once this capability is enabled. This capability enables upcoming features such as automatic component style hot module replacement (HMR) and development server deferred stylesheet processing.

To provide support for HMR of inline component styles (`styles` decorator field), the AOT compiler will now use the resource host transformation API with the Angular CLI to provide external runtime stylesheet URLs when the `externalRuntimeStyles` compiler option is enabled. This allows both a component's file-based and inline styles to be available for HMR when used with a compatible development server such as with the Angular CLI. No behavioral change is present if the `externalRuntimeStyles` option is not enabled or the resource host transformation API is not used. An `order` numeric field is also added to the transformation API which allows consumers such as the Angular CLI to create identifiers for each inline style in a specific containing file.

Runtime: https://github.com/angular/angular/pull/57922

clydin committed:
refactor(compiler): support external runtime component styles for file-based stylesheets

The AOT compiler now has the capability to handle component stylesheet files as
external runtime files. External runtime files are stylesheets that are not embedded
within the component code at build time. Instead a URL path is emitted within a component's
metadata. When combined with separate updates to the shared style host and DOM renderer,
this will allow these stylesheet files to be fetched and processed by a development
server on-demand. This behavior is controlled by an internal compiler option `externalRuntimeStyles`.
The Angular CLI development server will also be updated to provide the serving functionality
once this capability is enabled. This capability enables upcoming features such as automatic
component style hot module replacement (HMR) and development server deferred stylesheet processing.
The current implementation does not affect the behavior of inline styles. Only the
behavior of stylesheet files referenced via component properties `styleUrl`/`styleUrls`
and relative template `link` elements are changed by enabling the internal option.

clydin committed:
refactor(compiler-cli): add inline stylesheet external runtime style transformation support

To provide support for HMR of inline component styles (`styles` decorator field), the AOT
compiler will now use the resource host transformation API with the Angular CLI to provide
external runtime stylesheet URLs when the `externalRuntimeStyles` compiler option is enabled.
This allows both a component's file-based and inline styles to be available for HMR when used
with a compatible development server such as with the Angular CLI. No behavioral change is
present if the `externalRuntimeStyles` option is not enabled or the resource host transformation
API is not used.
An `order` numeric field is also added to the transformation API which allows consumers such as
the Angular CLI to create identifiers for each inline style in a specific containing file.

JoostK commented on packages/compiler-cli/src/ngtsc/core/api/src/options.ts:
```diff
@@ -96,6 +96,16 @@ export interface InternalOptions {
   _enableLetSyntax?: boolean;
 
+  /**
+   * Enables the use of link elements for component styleUrls instead of inlining the file
```
nit

```suggestion
   * Enables the use of `link` elements for component styleUrls instead of inlining the file
```

or


```suggestion
   * Enables the use of `<link>` elements for component styleUrls instead of inlining the file
```

JoostK commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -297,7 +305,9 @@ export function compileComponentFromMetadata(
-  } else if (meta.encapsulation === core.ViewEncapsulation.Emulated) {
+  }
+
+  if (!hasStyles && meta.encapsulation === core.ViewEncapsulation.Emulated) {
```
Would be good to have this change covered by a test as well, as it's quite subtle.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/test/component_spec.ts:
```diff
@@ -359,6 +366,269 @@ runInEachFileSystem(() => {
+      const {analysis} = handler.analyze(TestCmp, detected.metadata);
+      expect(analysis?.resources.styles.size).toBe(1);
+      expect(analysis?.meta.externalStyles).toEqual([]);
+      expect(analysis?.meta.styles).toEqual(jasmine.arrayWithExactContents(['.abc {}']));
```
```suggestion
      expect(analysis?.meta.styles).toEqual(['.abc {}']);
```

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/test/component_spec.ts:
```diff
@@ -359,6 +366,269 @@ runInEachFileSystem(() => {
+      const {analysis} = handler.analyze(TestCmp, detected.metadata);
+      expect(analysis?.resources.styles.size).toBe(1);
+      expect(analysis?.meta.externalStyles).toEqual(['abc/myInlineStyle.css']);
+      expect(analysis?.meta.styles).toEqual(jasmine.arrayWithExactContents([]));
```
```suggestion
      expect(analysis?.meta.styles).toEqual([]);
```

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/component/test/component_spec.ts:
```diff
@@ -359,6 +366,269 @@ runInEachFileSystem(() => {
+      const {analysis} = handler.analyze(TestCmp, detected.metadata);
+      expect(analysis?.resources.styles.size).toBe(1);
+      expect(analysis?.meta.externalStyles).toEqual([]);
+      expect(analysis?.meta.styles).toEqual(jasmine.arrayWithExactContents(['.abc {}']));
```
```suggestion
      expect(analysis?.meta.styles).toEqual(['.abc {}']);
```

JoostK commented on packages/compiler-cli/src/ngtsc/core/api/src/options.ts:
```diff
@@ -96,6 +96,16 @@ export interface InternalOptions {
+   *
+   * @internal
+   */
+  externalRuntimeStyles?: boolean;
```
Is the rationale behind the "runtime" naming that this causes styles to be requested ad-hoc at runtime upon first render of a component?

clydin replied:
Yes, the styles only become true external styles as per HTML at runtime. The primary goal was to avoid confusion with the existing `styleUrl`/`styleUrls` field where some may already consider them "external" styles.

clydin commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -379,19 +381,22 @@ export class ComponentDecoratorHandler
       }
       if (templateInfo.templateStyles) {
+        // For inline templates, the style element order includes the decorator styles if any
+        const orderBase = templateInfo.templateUrl == undefined ? (rawStyles?.length ?? 0) : 0;
```
Reduced the complexity of the logic here.  Also realized the component class name would be useful to uniquely identify the inline styles for cases where two components are in the same file.

clydin committed:
test(compiler-cli): add compliance tests for external runtime styles

To ensure that the external runtime style component feature is correctly
emitted by the Angular compiler, compliance tests have been added for
file-based component styles. Additionally, the partial golden generator
has been updated to work with file-based component styles.
